### PR TITLE
Que el hotel pueda decidir qué servicios sí deberían generar encuestas

### PR DESCRIPTION
**Pull request by FranzLuepke**
[CONT-38](https://app.clickup.com/t/3072919/CONT-38)

The configuration page for polls didn't exist, so a new configuration page was created, after the guest home organization.

The new polls page is shown below, with all the polls deactivated and no message:
Encuestas_0

The poll messages can be created, edited and the polls can be activated or deactivated:
Encuestas_Hotel_1

In this example, Room poll was activated:
Encuestas_Hotel_2

After the request is finished by the employer, the custom poll message appears (only if the hotel administrator had enabled the poll message to appear):
Habitación

### CODE REVIEW

File: apps/hotel/src/preferences/pages/polls/polls.component.ts [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+        this.addZones();
+      });
+    this.authService.$hotel
+      .pipe(switchMap((hotel) => this.ecommerceService.getEcommerceByHotelUid(hotel.uid)))
+      .subscribe((data) => {
+        this.ecommerceEntity = data;
+        this.addEcommerceZones();
+      });
+    this.authService.$hotel
+      .pipe(switchMap((hotel) => this.pollsService.getPolls(this.hotel.uid)))
```
Comment (ruedagato) [inline apps/hotel/src/preferences/pages/polls/polls.component.ts:68]:
```suggestion
      .pipe(first(), switchMap((hotel) => this.pollsService.getPolls(this.hotel.uid)))
```

File: apps/hotel/src/preferences/pages/polls/polls.component.ts [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+  }
+
+  ngOnDestroy(): void {
+    if (this.unsubscribe) {
+      this.unsubscribe.unsubscribe();
+    }
+  }
+
+  async goToHome(): Promise<void> {
+    const loader = this.messagesService.showLoader();
```
Comment (ruedagato) [inline apps/hotel/src/preferences/pages/polls/polls.component.ts:81]: this loader is not necessary if there isn't async logic

File: libs/core/src/lib/services/polls/polls.service.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+export class PollsService {
+  constructor(private afDb: AngularFireDatabase, private translate: TranslateService) {}
+
+  createPoll(
+    message: string,
+    status: boolean,
+    hotel: HotelEntity,
+    specialZoneHotelEntity: SpecialZoneHotelEntity,
+  ) {
+    console.log('PollsService: createPoll');
```
Comment (ruedagato) [inline libs/core/src/lib/services/polls/polls.service.ts:23]: remove logs

File: libs/core/src/lib/services/polls/polls.service.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  getPoll(hotelUid: string, zoneId: string): Observable<Polls | null> {
+    return this.afDb.object<Polls>(`specialZoneGuest/${hotelUid}/${zoneId}`).valueChanges();
+  }
+
+  getPolls(hotelUid: string): Observable<{ [key: string]: Polls }> {
+    return this.afDb.object<{ [key: string]: Polls }>(`polls/${hotelUid}/`).valueChanges();
+  }
+
+  updatePoll(poll: Polls, msg?: string) {
+    console.log(poll);
```
Comment (ruedagato) [inline libs/core/src/lib/services/polls/polls.service.ts:57]: remove log

File: apps/hotel/src/preferences/pages/polls/polls.component.ts [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+import { ZoneService } from '@contler/hotel/zone/services/zone.service';
+import { EcommerceEntity } from '@contler/entity/ecommerce.entity';
+
+@Component({
+  selector: 'contler-polls',
+  templateUrl: './polls.component.html',
+  styleUrls: ['./polls.component.scss'],
+})
+export class PollsComponent implements OnInit, OnDestroy {
+  hotel: HotelEntity;
```
[the remark this answers is not in the document]
Comment (FranzLuepke) [reply apps/hotel/src/preferences/pages/polls/polls.component.ts:26]: I'm now using hotel variable.

File: apps/hotel/src/preferences/pages/polls/polls.component.ts [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+
+  @HostListener('window:beforeunload', ['$event'])
+  canDeactivate(): boolean {
+    return !this.hasChanges;
+  }
+
+  ngOnInit(): void {
+    this.unsubscribe = this.authService.$hotel
+      .pipe(
+        tap((hotel) => (this.hotel = hotel)),
```
[the remark this answers is not in the document]
Comment (FranzLuepke) [reply apps/hotel/src/preferences/pages/polls/polls.component.ts:54]: Using hotel variable.

File: apps/hotel/src/preferences/pages/polls/polls.component.ts (hunk 2 above)
Comment (FranzLuepke) [reply apps/hotel/src/preferences/pages/polls/polls.component.ts]: Now goToHome function doesn't exist.

File: apps/guest/src/home/components/modal-booking-qualify/modal-booking-qualify.component.ts [hunk 7]
```diff
@@ -12,15 +16,27 @@ import { TEXT_RATE } from '../const/qualify.const';
 export class ModalBookingQualifyComponent implements OnInit {
   value: number | undefined;
   readonly textRate = TEXT_RATE;
+  readonly pollServiceConstant = POOL_SERVICE_CONSTANT;
   load = false;
+  polls: { [key: string]: Polls } = {};
```
Comment (ruedagato) [inline apps/guest/src/home/components/modal-booking-qualify/modal-booking-qualify.component.ts:21]:
```suggestion
  polls: Polls;
```

File: apps/guest/src/home/components/modal-booking-qualify/modal-booking-qualify.component.ts [hunk 8]
```diff
@@ -12,15 +16,27 @@ import { TEXT_RATE } from '../const/qualify.const';
...
+    public pollsService: PollsService,
+    private guestService: GuestService,
   ) {}
 
-  ngOnInit() {}
+  ngOnInit() {
+    this.guestService.$hotel.pipe(first()).subscribe((hotel) => {
+      this.pollsService.getPoll(hotel.uid, this.pollServiceConstant.reservations).subscribe((data) => {
+        this.service = this.pollServiceConstant.reservations;
+        this.polls = { [this.pollServiceConstant.reservations]: data };
```
Comment (ruedagato) [inline apps/guest/src/home/components/modal-booking-qualify/modal-booking-qualify.component.ts:36]:
```suggestion
        this.polls = data;
```

File: apps/hotel/src/preferences/pages/guest-home/guest-home.component.ts [hunk 9]
```diff
@@ -65,19 +65,19 @@ export class GuestHomeComponent implements OnInit, OnDestroy {
...
         this.hotel.uid,
         this.specialZoneGuest,
       );
       this.messagesService.closeLoader(loader);
-      await this.router.navigate(['home']);
     } catch (e) {
       this.messagesService.closeLoader(loader);
       this.messagesService.showServerError();
     }
+    await this.router.navigate(['preferences', 'polls']);
```
Comment (ruedagato) [inline apps/hotel/src/preferences/pages/guest-home/guest-home.component.ts:80]: if there is an error on line 71 it shouldn't happen, move this line after line 75

File: apps/hotel/src/preferences/pages/polls/edit-polls-guest/edit-polls-guest.component.scss [hunk 10]
```diff
@@ -0,0 +1,16 @@
+@import 'variables';
+@import 'mixin';
+
+.title {
+  @include text($text-3, 20px, 500);
```
Comment (ruedagato) [inline apps/hotel/src/preferences/pages/polls/edit-polls-guest/edit-polls-guest.component.scss:5]:
```suggestion
  @include text($text-3, #55413a, 500);
```

File: apps/hotel/src/preferences/pages/polls/edit-polls-guest/edit-polls-guest.component.scss [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+  letter-spacing: 0.25px;
+  text-align: center;
+  color: #55413a;
```
Comment (ruedagato) [inline apps/hotel/src/preferences/pages/polls/edit-polls-guest/edit-polls-guest.component.scss:8]: delete color

File: apps/hotel/src/preferences/pages/polls/polls.component.scss [hunk 12]
```diff
@@ -0,0 +1,46 @@
+@import 'variables';
+@import 'mixin';
+
+.title {
+  @include text($text-3, 20px, 500);
```
Comment (ruedagato) [inline apps/hotel/src/preferences/pages/polls/polls.component.scss:5]: same here

File: apps/hotel/src/preferences/pages/polls/polls.component.ts [hunk 13]
```diff
@@ -0,0 +1,198 @@
+import { Component, HostListener, OnDestroy, OnInit } from '@angular/core';
+import { map, switchMap, tap, filter, first } from 'rxjs/operators';
+import { AuthService } from '@contler/hotel/services/auth.service';
+import { Observable, Subscription, from } from 'rxjs';
+import { MatSlideToggleChange } from '@angular/material/slide-toggle';
+import { HotelEntity } from '@contler/entity';
+import { PollsService, EcommerceService } from '@contler/core'; //SpecialZoneGuestService
```
Comment (ruedagato) [inline apps/hotel/src/preferences/pages/polls/polls.component.ts:7]:
```suggestion
import { PollsService, EcommerceService } from '@contler/core';
```

File: apps/hotel/src/preferences/pages/polls/polls.component.ts [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+  }
+
+  ngOnDestroy(): void {
+    if (this.unsubscribe) {
+      this.unsubscribe.unsubscribe();
+    }
+  }
+
+  async goToHome(): Promise<void> {
+    try {
```
Comment (ruedagato) [inline apps/hotel/src/preferences/pages/polls/polls.component.ts:95]: try is not necessary

**Review by ruedagato — APPROVED** [review]

**Pull request merged**